### PR TITLE
[node] remove lib: "dom"; add global Event and EventTarget, fix other global types

### PR DESCRIPTION
**Pull request by thw0rted**
Please fill in this template.

- [X] Use a meaningful title for the pull request. Include the name of the package modified.
- [X] Test the change in your own code. (Compile and run.)
- [ ] [Add or edit tests](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#my-package-teststs) to reflect the change.  (Existing tests should be sufficient; behavior should not change.)
- [X] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [X] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [X] [Run `npm test <package to test>`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#running-tests).

If changing an existing definition:
- [X] Provide a URL to documentation or source code which provides context for the suggested changes: https://nodejs.org/dist/latest-v16.x/docs/api/buffer.html#class-blob

----

The previous version polluted the global namespace with a `Blob` interface which merged (incorrectly) with the DOM Blob definition provided by ts-lib.  My update uses the correct Node (experimental) Blob definition, which already existed under `node:buffer`.

I also took this opportunity to resolve an outstanding TODO comment, where `Blob#stream()` returned unknown because at the time it was originally added, the Node (experiemental) WHATWG / W3C compatible ReadableStream implementation didn't have types defined yet.  They have since been created (under `node:stream/web`), so that method now returns the correct stream type.

### CODE REVIEW

**Comment by thw0rted** [comment]
If anybody has advice for how to test that Node types do *not* pollute the global namespace with a Blob interface, I'd love to add one.  Failing that, behavior should remain the same as it was before.

**Comment by thw0rted** [comment]
This PR is going to require a bigger change than I originally expected.  The problem is that the package `tsconfig` is using `lib: ["dom"]`.  I changed that to `"es6"`, which broke a few things. As a result, I have added a full interface declaration for Node's DOM-compatibility Event and related types, under a fictitious module (`node:dom-events`).

If they were exported at the global level -- which, to be clear, they should be! -- they would break every project that uses both `@types/node` and `lib.dom.d.ts`, which is of course how we got into this mess in the first place.  There was a long conversation about how to handle this [in another issue](https://github.com/DefinitelyTyped/DefinitelyTyped/issues/34960) but I don't think we really got a good answer.  (*Consuming* code can do some trickery with conditional types evaluated against `globalThis` but I don't see how to fix it for use within the types-package.)

I think we may need smarter Typescripters than myself to take a look at this and suggest a way forward.  Are there other fundamentally-incompatible global types that differ between Node and the DOM?  If so, how have they been handled?  If not... now what?

**Comment by thw0rted** [comment]
The CI build keeps failing but I don't understand how the broken packages ever worked.  Two of them depend on `@types/mongodb` which was removed last year (Mongo now ships their own types), and the old package is flat-out broken.  One of them has `$ExpectType "one" | "two" | "three"` but [that syntax never worked](https://github.com/microsoft/dtslint/issues/191) as far as I can tell.  Any help with resolving the CI problems would be much appreciated.

**Comment by thw0rted** [comment]
The change I just pushed exposes Event and EventTarget globally via the same mechanism used in https://github.com/DefinitelyTyped/DefinitelyTyped/pull/58277.  I'm not sure how to test and make sure this doesn't explode real-world users...

**Comment by thw0rted** [comment]
In hindsight, this PR has grown in scope, maybe too much.  I just created https://github.com/thw0rted/DefinitelyTyped/tree/node-minus-dom which is a tiny patch that only changes the tsconfig `lib` setting.  There were three places that broke:

* one reference to `performance` was simply mistaken; Node does not expose that as a global, you're supposed to import it from `perf_hooks`
* one reference to global `WebAssembly` which has not been implemented yet.  All references in `WASI` just have a TODO comment, so I commented out the reference and left a copy of the same TODO comment in the relevant test.
* One reference to global EventTarget.  I put in a workaround that refers to the (unexported) "DOMEventTarget" interface; at some point actually making a global EventTarget would be better.

If the reviewer(s) would prefer I can file a separate PR with just that change, then maybe a separate PR just for global Event/EventTarget and another one for the Blob/Stream stuff?  Not sure what would represent less overall work/complexity/controversy.

**Comment by thw0rted** [comment]
I rebased to the latest master and now for some reason the workaround to conditionally expose Event as a global doesn't seem to work in TS 4.7.  (When I initially submitted this, I don't think 4.7 was being used for tests?)

I'm a bit out of my depth trying to track this down.  As I understand it, the conditional works by checking if an `onmessage` global is defined -- which it is, in `lib.dom.d.ts`, even in 4.7 -- and re-declaring the existing (DOM) Event interface if so, or making a global alias for the Node DOM-like Event if not.  I can't figure out why that would work in 4.6 but not 4.7.

I think @antongolub and @peterblazejewicz may have been responsible for this construct, if you have a minute to see why it's not working here (anymore?) I'd really appreciate it.  `npm run test zui` reproduces the error, or pick any other package that uses `lib: ["dom"]`.

For now, I'll temporarily remove the global Event stuff and see if I can get the CI passing.

**Comment by antongolub** [comment]
@thw0rted,

1) https://github.com/antongolub/DefinitelyTyped/tree/thw0rted-fix-dom-events
2) What's `dom-events` module in nodejs? Could you provide a ref to its docs?

**Comment by thw0rted** [comment]
Thank you, I will take a look at your branch, I think that will help.

I commented upthread that the `node:dom-events` module is "fictitious". I created it to make it easier to reference the Node DOM-like Event and related interfaces, from e.g. test code, so that it will work whether or not the new interfaces are exposed at the global level.  It should contain exactly the interfaces defined in [the Events docs](https://nodejs.org/api/events.html).

I don't know if this is a bad idea -- I guess Node could decide to make a real `dom-events` module at some point? -- and would welcome feedback to that effect.  I was just trying to avoid having more module-level Typescript-only interfaces like the current "DOMEventTarget".  Basically, I thought it would be better to have a Typescript-only module whose interfaces accurately reflect the Node docs, rather than making up new names for interfaces inside a real module.

**Comment by antongolub** [comment]
1. I ran `dtslint types node` only
2.   `Types of property 'bubbles' are incompatible.
    Type 'boolean' is not assignable to type 'false'
     Types of property 'cancelBubble' are incompatible.
    Type 'boolean' is not assignable to type '() => void'.`
    and so on.
    
Now you need to smooth out the iface diff.

**Comment by thw0rted** [comment]
It looks like you're trying to reconcile the two interfaces with each other so that they can be merged without conflicts.  This isn't actually possible, which is why I was trying to use the workaround where only one or the other can exist globally, without merging.

I think I may have found other inconsistencies but the first that jumps out at me is `cancelBubble`, which is a [boolean property in the DOM](https://developer.mozilla.org/en-US/docs/Web/API/Event/cancelBubble) and [a method (`() => void`) in Node](https://nodejs.org/api/events.html#class-event). 

I think the safest thing is if we can make sure that users with `@types/node` and no `lib.dom.d.ts` see Node's `Event` as a global, but users with both libraries only see the DOM `Event`, and have to import the Node version from somewhere else (like my "fictional" module).

**Comment by thw0rted** [comment]
OK so, at the risk of burning another hour of CI time (sorry!), I'm pushing a commit that allows the two interfaces to merge again.  One change is basically trivial (typing `bubbles` as `false`) and one says in the docs it should not be used (`cancelBubble`).  I thought there were others but it looks like I may have been wrong, so these can peacefully coexist for now.

**Comment by amcasey** [comment]
@thw0rted Should this be marked as draft?  It looks like you're still working through some issues.

**Comment by thw0rted** [comment]
Hi @amcasey, I haven't looked at this in a few weeks but I think it worked the way I left it.  I know this is a big change and I'm not sure what the process is supposed to be, especially since so many people use this package.  I wouldn't say it's a draft, but it could probably use some discussion time, I'm just not sure who should be involved.

**Review by sheetalkamat — CHANGES_REQUESTED** [review]
This is a big change and it needs reviews from authors

File: types/node/test/stream.ts
Comment (LinusU) [inline types/node/test/stream.ts:17]: Node.js 18.0.0 has a global Blob for me 🤔 

Screenshot 2022-07-07 at 18 09 16

Comment (thw0rted) [reply types/node/test/stream.ts:17]: I can't find any docs for this.  Do you know when it was introduced?  I'm still on an older LTS (14.7 I think? Don't update that often.) which definitely does not have a global Blob.

Can you please run `Blob === buffer.Blob` for me and see what it says?  May need to `import buffer from 'buffer'` first.

Comment (thw0rted) [reply types/node/test/stream.ts:17]: So, it looks like as of Node v18, Blob is moved/aliased into the global namespace intentionally, which negates a big part of the motivation for this PR. The problem is, Node's Blob must merge correctly with the one in lib-dom, but can't reference it directly unless/until we get something like [this feature](https://github.com/microsoft/TypeScript-DOM-lib-generator/issues/1207), which hasn't seen any movement in a while.

I'm not sure how to proceed.  The motivating case is still an issue -- projects that include both `@types/node` and `lib: ['dom']` can still see buggy behavior.  Maybe some of the core `@types/node` authors could weigh in here?

Comment (LinusU) [reply types/node/test/stream.ts:17]: I should add that I think that this PR is a really good idea (with my limited understanding) since most dom apis are _not_ available in Node.js, and having it include `dom` isn't super...

Comment (LinusU) [reply types/node/test/stream.ts:17]: Screenshot 2022-07-08 at 14 58 37

**Comment by fredericvilcot** [comment]
Hey guys, is there any progress on this PR or any other possible fix for the conflict between `dom` and `nodejs` types for the `Blob`?
Pointing the `Blob` interface from the `dom` lib is not enough?
How can we help with this resolution?
Many thanks

**Comment by thw0rted** [comment]
@fredericvilcot I really don't know, to my knowledge this might be the first time Node has placed an interface at the global scope that directly conflicts with a DOM interface of the same name.  TS has kicked around some potential solutions but there didn't seem to be any urgency since projects that incorporated both `@types/node` and `lib-dom` mostly didn't manifest any obvious bugs.  I think, especially with the global scope of `Blob` etc starting with Node v18, we're going to see more bugs more often, so maybe that will be the push we need to get the TS team to take action.  (I'm still not sure what that action should look like, though.)

**Comment by fredericvilcot** [comment]
Thank you @thw0rted for your quick return :)
While waiting for a possible fix on the TS side, is there a workaround to overcome this issue without having to re-cast the properties of `Blob` in `any` or other?
I can't remove the `@types/node` in my project because other libraries need them to work.. and I'm evolving in a `dom` environment so there are type conflicts...

**Comment by thw0rted** [comment]
Sorry, I'm not a maintainer of these types and I don't have any particular insight into the Typescript team, I just saw a problem and (before Node put `Blob` in the global namespace!) thought maybe I could do something about it.  Now that a very commonly used package (Node) has types that directly conflict with the most commonly used core library (lib-dom), I really think the issue needs to be elevated to the core TS team.  In the interim, I think the least-damaging thing would be to update Node's typings to pretend that global Blob does not exist, and constrain NodeBlob to the `buffer` library.  It's bad, but seems less bad to me than the alternative.

**Comment by peterblazejewicz** [comment]
@thw0rted I can work with you doing those steps, let's remove globals. THe docs favor namespaced named imports anyway. But one symbol at time (say Blob first), let's change and wait for possible fallout

**Comment by fredericvilcot** [comment]
@peterblazejewicz if you need any help, don't hesitate😌

**Comment by thw0rted** [comment]
I merged the current master and updated a few broken tests.  I also tried to conditionally expose the `Blob` constructor globally using a construct copied from `URLSearchParams`:

```ts
        var Blob: typeof globalThis extends {
            onmessage: any;
            Blob: infer Blob;
        }
            ? Blob
            : typeof import('node:buffer').Blob;
```

I think this construct actually works well, but one wrinkle I had forgotten was that this only exposes the value-side, not the type-side, of `Blob`.  `URLSearchParams` has [an interface declaration](https://github.com/thw0rted/DefinitelyTyped/blob/node-blob-fix/types/node/url.d.ts#L865) as well, but this only works because that interface does not conflict with the one in `lib-dom`.  If I tried e.g. `interface Blob extends _Blob {}` I would get errors any time a library included both `@types/node` and `lib: ["dom"]`.  (Try `npm run test atom` for an example.)

As far as I can tell, it is not possible to conditionally expose the type-side.  There is no such thing as a conditional interface augmentation, and declaring it as a mapped / conditional `type` (`type Blob = typeof globalThis extends {....} ? never : _Blob`) creates a duplicate identifier when both libs are included.  I'd love a third option here, and it's certainly possible that my type-fu is just too weak -- help would be welcome.

Barring that: what if we take out all the global alias *interfaces* (only), and leave in the constructors?  So, `new Blob(...)` would work, but `const b: Blob = getBlob(...)` would not.  You can still write `InstanceType<typeof Blob>` to get access to the type-side at global scope, or just import from the scoped package.  How does that sound?

File: types/node/util.d.ts
Comment (peterblazejewicz) [inline types/node/util.d.ts:1116]:
```suggestion
    global {
        /**
         * `TextEncoder` class is a global reference for `require('util').TextEncoder`
         * https://nodejs.org/api/globals.html#textencoder
         * @since v11.0.0
         */
        var TextEncoder: {
            prototype: TextEncoder;
            new (): TextEncoder;
        };
    }
```
that will remove current errors. Is that symbol export you'd expect here?

Comment (thw0rted) [reply types/node/util.d.ts]: Sorry I was still running tests last night so I didn't get to commit my fixes.  This should be OBE with the commit I pushed -- TextEncoder is just one of the DOM-like globals that have been added in the last couple of Node versions.  The updated branch should expose all the non-experimental global constructors conditionally (but none of the matching interfaces).

File: types/node/test/stream.ts
Comment (thw0rted) [reply types/node/test/stream.ts]: Per the discussion below I'm suggesting globally exposing the Blob constructor, which we can do conditionally, but not the Blob type, which we can't.  Resolving this, since it's part of the larger discussion below.

File: types/node/test/crypto.ts
Comment (peterblazejewicz) [inline types/node/test/crypto.ts:1392]: 👍🏻

File: types/node/events.d.ts
Comment (peterblazejewicz) [inline types/node/events.d.ts:650]: is this specific bit (dom-events) required for this PR? that one will expose `node:dom-events` for autoimports, failing at runtime (unknown builtin module)

Comment (thw0rted) [reply types/node/events.d.ts:650]: The plan was, if the module's exports live entirely in type-space, the `import` statements should never make it into the emit.  Looking back over it, that isn't actually the case -- the fictional module declares `const Event` and `const EventTarget`, but of course `import {EventTarget} from "node:dom-events"; new EventTarget();` would fail.  I'll try to come up with another approach.

Also: I just checked and `NodeEventTarget` is defined in the real `events` module... but *never referenced anywhere else* -- no other Node API uses this interface.  I guess the idea is for third party libraries to inherit from it?  But the interface in the current code is missing most of the properties and methods.  Will look at moving this into the real module.

Comment (thw0rted) [reply types/node/events.d.ts]: Update on the NodeEventTarget situation: that class appears in the docs but [is not actually available at runtime](https://github.com/nodejs/node/issues/39903), so I just commented the whole thing out.  It should be easy to add it back if the linked issue gets resolved.

Comment (thw0rted) [reply types/node/events.d.ts]: The changes I'm pushing now should address the events-module problems, I think.  I realized that 1) if the module is not intended for end-users to consume, it should be OK to make intermediate / alias types, and 2) I can use the same `typeof globalThis extends ....` construct to conditionally define an alias, then declare the global interface augmentation as `extends __Alias`.  As a result, there should be a global `Event` type *and* constructor, and both "sides" should reflect either the Node version or the DOM one (but never both).

I'm not sure this is "done", though.

First, it worked for Event and EventTarget because they are defined as interfaces on the type side with a `var Event` that defines the constructor side.  It wouldn't work the same way with class declarations, though in fairness I don't think many of the current globals are declared as `class`es.

Second, it necessarily involves creating extra modules and/or types that aren't intended for public use.  I've always been a little uncomfortable working with module definitions -- what makes it "ambient"?  When does this interface get exported even though I didn't say "export"?  How am I importing from a module, inside that same module? etc etc -- so I feel like if I just understood it a bit better, I could more effectively hide the "work in progress" types and whatnot.  (Suggestions are more than welcome!)

Comment (peterblazejewicz) [reply types/node/events.d.ts]: IMO it works now, nothing can be imported, so it impacts only global symbols

File: types/node/events.d.ts
Comment (rbuckton) [inline types/node/events.d.ts:688]: What else uses this aside from the commented section on line 42? Is it still needed? 

It also doesn't seem necessary to introduce a type-only pseudo-module when these definitions could be moved inside the `events` module and just not exported (i.e., add an `export` to each declaration in `events` along with an `export {};` to indicate that declarations in the module aren't exported by default.

Comment (thw0rted) [reply types/node/events.d.ts]: Not gonna lie, I only learned about `export {}` about a week ago.  As it turns out though, that isn't even necessary, because the `events` module uses a default-assignment export (`export = EventEmitter`) so everything else is already not exported.

It occurred to me though, the DOM-like Event(Target) is only used globally, never exported from any module, and contains no references to anything from the `event` module, so there's no reason for that code to live in this file.  I moved it to its own file in the latest commit.

File: types/node/index.d.ts
Comment (rbuckton) [inline types/node/index.d.ts:95]: I'm surprised this passes tests given that `dom-events` should be parsed as a module.

Comment (thw0rted) [reply types/node/index.d.ts:95]: Ah, well, I may be showing a lot of ignorance in this PR but at least I'm learning a lot too.

Is the opposite of "parsed as a module" supposed to be "parsed as a script"?  And all the other `reference`d files parse as a script because at the top level they only `declare module 'xyz'`?  (Or in the case of `globals.d.ts`, declare types that will be added to the global scope.)

Should I refactor the DOM event declarations to parse as a script?  (Is that possible without exporting all the intermediate types?)  If not, would it be more idiomatically correct to use a side-effect `import` statement instead?  I was just aping the syntax used to incorporate the other files, not realizing how they were different.

**Comment by sheetalkamat** [comment]
@typescript-bot this still has merge conflicts

**Comment by sandersn** [comment]
There are no merge conflicts anymore, but I don't see signoffs from any of the reviewers either. @LinusU, @rbuckton, @peterblazejewicz do you want to approve this now?

**Comment by thw0rted** [comment]
I don't want to derail this if somebody is actually planning to merge it, but at least a note for potential future work: https://github.com/DefinitelyTyped/DefinitelyTyped/issues/60924#issuecomment-1246624868 points out that Node's `fetch` is actually just taken from `undici` (an official Node project), which ships its own types, including their interpretation of all the WHATWG streams and whatnot.  Any types currently living under `@types/node` that also exist under `undici` should be removed and pulled from there instead by reference.  (If Node wants to maintain their own types, that sounds *great* to me...)

File: types/node/stream/consumers.d.ts
Comment (rbuckton) [inline types/node/stream/consumers.d.ts:2]: Is the import alias necessary here?

File: types/node/url.d.ts
Comment (rbuckton) [inline types/node/url.d.ts:11]: Is the impot alias necessary here?

**Review by rbuckton — APPROVED** [review]
A few minor questions, but otherwise this looks fine.

File: types/node/stream/consumers.d.ts
Comment (thw0rted) [reply types/node/stream/consumers.d.ts:2]: It isn't but in my own projects I've made a habit of aliasing anything that shadows a global type name, just to be sure there is never the possibility of confusion.  Happy to change it back if that's a style preference for the project or whatever.

Comment (rbuckton) [reply types/node/stream/consumers.d.ts:2]: I'm not concerned. you can go ahead and ask the bot to merge.

**Comment by thw0rted** [comment]
Ready to merge

🤞🤞🤞🤞🤞

**Comment by thw0rted** [comment]
Ready to merge

**Comment by AviVahl** [comment]
v16.11.65 caused a regression on my end. Can be seen in https://github.com/wixplosives/pleb, by cloning, removing the lock file, and running `npm it`:
```
node_modules/@types/node/ts4.8/stream.d.ts:842:39 - error TS2304: Cannot find name 'Blob'.

842             static from(src: Stream | Blob | ArrayBuffer | string | Iterable<any> | AsyncIterable<any> | AsyncGeneratorFunction | Promise<any> | Object): Duplex;
                                          ~~~~

node_modules/@types/node/ts4.8/stream.d.ts:842:39 - error TS2304: Cannot find name 'Blob'.

842             static from(src: Stream | Blob | ArrayBuffer | string | Iterable<any> | AsyncIterable<any> | AsyncGeneratorFunction | Promise<any> | Object): Duplex;
```

Can be worked around by adding "dom" to tsconfig->lib, which kinda beats the purpose of this PR.

**Comment by Jack-Works** [comment]
This PR is breaking Vue 3. I'll draft issue in both Vue repo and DefinitelyTyped repo soon.

**Comment by Jack-Works** [comment]
See https://github.com/DefinitelyTyped/DefinitelyTyped/discussions/62759 and https://github.com/vuejs/core/issues/6899

**Pull request merged**
